### ISSUE TITLE
mpi4py Hello World fails

### ISSUE THREAD

**Issue by artofbeinghuman** [aws/aws-parallelcluster#1187]
**Environment:**
 - aws-parallelcluster-2.4.0
 - OS: ubuntu1604
 - Scheduler: SGE
 - Master instance type: c5n.large
 - Compute instance type: c5n.large

**What I want to do:**
I want to scale my python MPI script from one c5 instance to multiple instances. I fail however to do a simple mpi4py hello world example.

**How I am trying to do it:**

locally: `~/.parallelcluster/config`:
```
[aws]
aws_region_name = us-east-2

[global]
cluster_template = evo
update_check = true
sanity_check = true

[aliases]
ssh = ssh -i ~/.ssh/**.pem {CFN_USER}@{MASTER_IP} {ARGS}

[vpc public]
master_subnet_id = subnet-***
vpc_id = vpc-****

[scaling custom]
scaledown_idletime = 15

[cluster evo]
vpc_settings = public
key_name = **
base_os = ubuntu1604
scaling_settings = custom
initial_queue_size = 3
max_queue_size = 10
maintain_initial_size = true
scheduler = sge
cluster_type = spot
compute_instance_type = c5n.large
master_instance_type = c5n.large
placement_group = DYNAMIC
placement = compute 
```

I log into the cluster using `pcluster ssh mycluster`. There I execute a setup file which is close to what I used to setup my single AWS instance. It is supposed to update the system, install some libraries and then setup my python 3.7 virtualenv with the appropriate pip packages. By putting the python virtualenv into the /shared directory, I hoped to be able to use the same setup for all the instances. Is this wrong? Do I have to give some sort of post_install script that sets all of this up for every single instance (compute and master)?

On cluster `setup.sh`:
```
sudo apt update && sudo apt-get -y upgrade &&
sudo apt --assume-yes install tmux bc &&
sudo apt update && sudo apt-get -y upgrade &&

cd /shared &&
mkdir -p /usr/src &&
cd /usr/src &&
sudo apt --assume-yes install python3-tk python3-pip &&
sudo apt --assume-yes install build-essential checkinstall &&
sudo apt --assume-yes install libreadline-gplv2-dev libncursesw5-dev libssl-dev libsqlite3-dev tk-dev libgdbm-dev libc6-dev libbz2-dev libffi-dev &&
sudo apt install -y python3-dev zlib1g-dev libjpeg-dev cmake swig python-pyglet python3-opengl libboost-all-dev libsdl2-dev libosmesa6-dev patchelf ffmpeg xvfb &&
sudo wget https://www.python.org/ftp/python/3.7.2/Python-3.7.2.tgz &&
sudo tar xzf Python-3.7.2.tgz &&
cd Python-3.7.2 &&
sudo ./configure --enable-optimizations &&
sudo make altinstall &&

# setup virtualenv with python libraries
cd /shared &&
pip3.7 install --user virtualenv &&
python3.7 -m virtualenv --python=python3.7 GymEnv &&
source /shared/GymEnv/bin/activate &&
pip install --upgrade pip &&
pip install numpy torch torchvision ipython matplotlib && 
pip install mpi4py baselines atari_py gym[atari] Pillow click &&
cd /shared
```

This script runs fine, so I setup my python MPI hello world example and my job file:
`test.py`:
```
from mpi4py import MPI

def func():
    comm = MPI.COMM_WORLD
    rank = comm.rank
    print("Hello World from rank {}\n".format(rank))

if __name__ == '__main__':
    func()
```
and `job.sh`:
```
#!/bin/bash
source /shared/GymEnv/bin/activate
mpirun /shared/GymEnv/bin/python test.py
```

This is run via: `qsub -cwd -V -pe mpi 6 job.sh`. I pass `6`, since I have three `c5n.large` compute instances, which have two vCPUs each.
I don't pass `-np 6` to mpirun in the job script because of [#815](https://github.com/aws/aws-parallelcluster/issues/815#issuecomment-453647971).

**The Errors:**
This produces two types of errors, here is the full output:
```
ubuntu@ip-172-31-0-16:/shared$ cat job.sh.e5
/var/spool/sge/ip-172-31-9-93/job_scripts/5: 2: /var/spool/sge/ip-172-31-9-93/job_scripts/5: source: not found
Could not find platform independent libraries <prefix>
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Could not find platform independent libraries <prefix>
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Could not find platform independent libraries <prefix>
Could not find platform independent libraries <prefix>
Fatal Python error: Fatal Python error: initfsencoding: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

initfsencoding: Unable to get the locale encoding
ModuleNotFoundError: Current thread 0xNo module named 'encodings'

Current thread 0x00002b39a891fe00 (most recent call first):
00002b8b38fcce00 (most recent call first):
Fatal Python error: initfsencoding: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00002b0daadebe00 (most recent call first):
Fatal Python error: initfsencoding: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00002b6245014e00 (most recent call first):
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: initfsencoding: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00002b07bb1f5e00 (most recent call first):
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: initfsencoding: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00002adc792e1e00 (most recent call first):
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 2 with PID 4769 on node ip-172-31-6-230 exited on signal 6 (Aborted).
--------------------------------------------------------------------------
```

**1.)** Trying to set the virtualenv, which I installed, as the python environment which should be used, fails because the `source` command doesn't work. Why?

**2.)** What I gather from different forum posts is that basically the python directory of my virtualenv cannot be properly found, and encodings is the first point where this becomes apparent. I have to tinker with the PYTHONHOME and PYTHONPATH variables, but have no idea how exactly.
I changed `job.sh` to:
```
#!/bin/bash
export PYTHONHOME=/shared/GymEnv/
export PYTHONPATH=/shared/GymEnv/bin
source /shared/GymEnv/activate
mpirun /shared/GymEnv/bin/python test.py
```
and this got rid of all the prefix errors, but the rest remains.

I would be very thankful for any advice on this problem. I think scaling from one c5d.18xlarge instance to 10 instances should be quite easy, so probably I am missing something big here. At the very least a python MPI hello World example should be doable.

Thanks, 
Marvin

**Bonus:**
As you can see, I set `maintain_initial_size = true` and `initial_queue_size = 3`, however, if I don't use the cluster, the three compute instances are shut down and have to be started anew for a new job. Apparently, I misunderstood what `maintain_initial_size` does, because I thought with setting this property to true, that there would always be `initial_queue_size` compute instances kept alive and running (docs say "If set to true, the Auto Scaling group will never have fewer members than the value of initial_queue_size").

**Comment by lukeseawalker** [comment]
Hi @artofbeinghuman,
first of all thank you for the very detailed report!

I have the suspect that you are not configuring/installing/setting up all the hosts in the same way, where "all the hosts" include both master and compute nodes.

As you were wondering, you should do the setup using a post_install script, see [here](https://docs.aws.amazon.com/parallelcluster/latest/ug/pre_post_install.html) for details, or even think to prepare the AMI in advance if the post install actions become many and require a considerable amount of time to be executed, see [here](https://docs.aws.amazon.com/parallelcluster/latest/ug/tutorials_02_ami_customization.html) for details.

My suspect is confirmed by the fact that I'm able to submit a simple python MPI hello world without all the virtualenv setup in this way:
* I have installed mpi4py in all the nodes (master and computes with) `sudo pip install mpi4py`
* prepared the python script 
```cat /shared/test.py
from mpi4py import MPI
import socket

def func():
    comm = MPI.COMM_WORLD
    rank = comm.rank
    print("Hello World from rank {0}, I'm {1}".format(rank, socket.gethostname()))

if __name__ == '__main__':
    func()
```
*  prepared the job script 
```
cat /shared/job.sh
#!/bin/bash
# Below the directives for SGE
#$ -pe mpi 14
#$ -R y
mpirun python /shared/test.py
```
* submitted the job with
```
qsub /shared/job.sh
```
finally I can check the output
```
cat job.sh.o1
Hello World from rank 0, I'm ip-172-31-65-50
Hello World from rank 8, I'm ip-172-31-77-144
Hello World from rank 13, I'm ip-172-31-77-144
Hello World from rank 1, I'm ip-172-31-65-50
Hello World from rank 3, I'm ip-172-31-65-50
Hello World from rank 4, I'm ip-172-31-65-50
Hello World from rank 5, I'm ip-172-31-65-50
Hello World from rank 7, I'm ip-172-31-65-50
Hello World from rank 9, I'm ip-172-31-77-144
Hello World from rank 10, I'm ip-172-31-77-144
Hello World from rank 11, I'm ip-172-31-77-144
Hello World from rank 12, I'm ip-172-31-77-144
Hello World from rank 6, I'm ip-172-31-65-50
Hello World from rank 2, I'm ip-172-31-65-50
```

Bonus answer. The point here is that you are using "spot" instance, so "If set to true, the Auto Scaling group will never have fewer members than the value of initial_queue_size" is always true for on-demand, but not for spot. I will take care of clarify that in the docs. You could double check in the AutoScaling Group console log the reason why the instances went terminated.

L

**Comment by artofbeinghuman** [comment]
Hi Luca,

thank you for your helpful response.
I created my AMI like you proposed and changed my cluster config accordingly:
```
[aws]
aws_region_name = us-east-2

[global]
cluster_template = evo
update_check = true
sanity_check = true

[aliases]
ssh = ssh -i ~/.ssh/**.pem ubuntu@{MASTER_IP} {ARGS}

[vpc public]
master_subnet_id = subnet-***
vpc_id = vpc-****

[scaling custom]
scaledown_idletime = 15

[cluster evo]
vpc_settings = public
key_name = **
#base_os = ubuntu1604
# Evo ParallelCluster AMI:
custom_ami = ami-0882fe6951fe9a741  
scaling_settings = custom
initial_queue_size = 3
max_queue_size = 10
maintain_initial_size = true
scheduler = sge
cluster_type = spot
compute_instance_type = c5n.large
master_instance_type = c5n.large
placement_group = DYNAMIC
placement = compute 
```

Now I run into another issue. Like I said, the spot instance shut down when unused, but they are spun back up, when a new job is submitted. However, when this happens, the new compute instances fail to "be connected" to the master apparently. They are clearly online in the AWS EC2 Console, but they fail to show up in `qhost` and thus also fail to execute my job. Amusingly, since the compute instances are not called upon to execute the job, they shut down agan shortly after and the job stays in qw:
```
ubuntu@ip-172-31-15-238:/shared$ qstat
job-ID  prior   name       user         state submit/start at     queue                          slots ja-task-ID 
-----------------------------------------------------------------------------------------------------------------
      1 0.55500 job.sh     ubuntu       qw    07/11/2019 13:08:24                                    6        

```


Someone already mentioned [here](https://github.com/aws/aws-parallelcluster/issues/86#issuecomment-196966385) that the issue might stem from custom AMIs but other than in the case of the linked issue, I do need to use a custom AMI.

Any idea on what to do? Of course I would greatly prefer to use spot instances instead of on-demand compute instances as I stand to save about 82% of costs due to using spot instances (c5n.18xlarge).

Thank you,
Marvin

**Comment by lukeseawalker** [comment]
To find the root cause please send `/var/log/nodewatcher` from the Compute node and `/var/log/jobwatcher` and `/var/log/sqswatcher` from the Master node.

The steps you did in order to create the AMI are good. Unfortunately, it's likely that the part "downloaded and compiled python3.7 and installed the necessary packages via pip3.7 install --user ..." is messing up with the python installation used by the nodewatcher, jobwatcher and sqswatcher daemons. The logs will prove that.

In the near future we are planning to setup a virtualenv where to run the ParallelCluster daemons.

**Comment by artofbeinghuman** [comment]
`/var/log/nodewatcher` from one of the compute nodes, right after startup and after `qsub job.sh`:
```2019-07-11 15:08:13,490 INFO [nodewatcher:main] nodewatcher startup
2019-07-11 15:08:13,491 INFO [nodewatcher:_get_config] Reading /etc/nodewatcher.cfg
2019-07-11 15:08:13,492 INFO [nodewatcher:_get_config] Configured parameters: region=us-east-2 scheduler=sge stack_name=parallelcluster-evcl scaledown_idletime=15 proxy=NONE
2019-07-11 15:08:14,047 INFO [nodewatcher:main] Instance id is i-070121e7d49dbb6fd, hostname is ip-172-31-6-200.us-east-2.compute.internal, instance type is c5n.large
2019-07-11 15:08:14,446 INFO [utils:get_asg_name] ASG parallelcluster-evcl-ComputeFleet-18TYEPZ0QJEPK found for the stack parallelcluster-evcl
2019-07-11 15:08:14,447 INFO [nodewatcher:_wait_for_stack_ready] Waiting for stack parallelcluster-evcl to be ready
2019-07-11 15:08:14,549 INFO [nodewatcher:_wait_for_stack_ready] Stack parallelcluster-evcl is in status: CREATE_IN_PROGRESS
2019-07-11 15:08:24,560 INFO [nodewatcher:_wait_for_stack_ready] Waiting for stack parallelcluster-evcl to be ready
2019-07-11 15:08:24,612 INFO [nodewatcher:_wait_for_stack_ready] Stack parallelcluster-evcl is in status: CREATE_IN_PROGRESS
2019-07-11 15:08:34,623 INFO [nodewatcher:_wait_for_stack_ready] Waiting for stack parallelcluster-evcl to be ready
2019-07-11 15:08:34,681 INFO [nodewatcher:_wait_for_stack_ready] Stack parallelcluster-evcl is in status: CREATE_COMPLETE
2019-07-11 15:08:34,702 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:08:34,702 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:08:44,721 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:08:44,721 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:08:54,735 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:08:54,735 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:09:04,746 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:09:04,747 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:09:14,758 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:09:14,758 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:09:24,768 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:09:24,769 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:09:34,780 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:09:34,780 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:09:44,791 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:09:44,791 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:09:54,802 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:09:54,802 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:10:04,813 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:10:04,814 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:10:14,833 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:10:14,833 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
2019-07-11 15:10:24,851 WARNING [sge:is_node_down] Node is not attached to scheduler. Reporting as down
2019-07-11 15:10:24,852 WARNING [nodewatcher:_poll_wait_for_node_ready] Node reported as down
```

On this compute node, `qstat` displays the job, but of course it is not executed.

On Master Node, 
`/var/log/jobwatcher`:
```
2019-07-11 15:05:36,079 INFO [jobwatcher:main] jobwatcher startup
2019-07-11 15:05:36,079 INFO [jobwatcher:_get_config] Reading /etc/jobwatcher.cfg
2019-07-11 15:05:36,080 INFO [jobwatcher:_get_config] Configured parameters: region=us-east-2 scheduler=sge stack_name=parallelcluster-evcl pcluster_dir=/opt/parallelcluster proxy=NONE
2019-07-11 15:05:36,390 WARNING [utils:get_asg_name] Unable to get ASG for stack parallelcluster-evcl
2019-07-11 15:05:56,736 WARNING [utils:get_asg_name] Unable to get ASG for stack parallelcluster-evcl
2019-07-11 15:06:36,977 INFO [utils:get_asg_name] ASG parallelcluster-evcl-ComputeFleet-18TYEPZ0QJEPK found for the stack parallelcluster-evcl
2019-07-11 15:06:38,216 INFO [jobwatcher:_poll_scheduler_status] Refreshing compute instance properties
2019-07-11 15:06:38,607 INFO [utils:_get_vcpus_by_instance_type] Instance c5n.large has 2 vcpus.
2019-07-11 15:06:38,608 INFO [utils:_read_cfnconfig] Reading /opt/parallelcluster/cfnconfig
2019-07-11 15:06:38,608 INFO [utils:get_instance_properties] Instance c5n.large will use number of vcpus as slots based on configuration.
2019-07-11 15:06:38,608 INFO [utils:get_instance_properties] Number of slots computed for instance c5n.large: 2
2019-07-11 15:06:38,816 INFO [utils:get_asg_settings] min/desired/max 0/3/10
2019-07-11 15:06:39,068 INFO [jobwatcher:_poll_scheduler_status] There are no pending jobs or the requirements on pending jobs cannot be satisfied. Noop.
2019-07-11 15:07:39,337 INFO [utils:get_asg_settings] min/desired/max 0/3/10
2019-07-11 15:07:39,343 INFO [jobwatcher:_poll_scheduler_status] There are no pending jobs or the requirements on pending jobs cannot be satisfied. Noop.
2019-07-11 15:08:39,560 INFO [utils:get_asg_settings] min/desired/max 0/3/10
2019-07-11 15:08:39,566 INFO [jobwatcher:_poll_scheduler_status] There are no pending jobs or the requirements on pending jobs cannot be satisfied. Noop.
2019-07-11 15:09:39,591 INFO [jobwatcher:_poll_scheduler_status] Refreshing compute instance properties
2019-07-11 15:09:39,890 INFO [utils:get_asg_settings] min/desired/max 0/3/10
2019-07-11 15:09:39,902 INFO [jobwatcher:_poll_scheduler_status] 3 nodes requested, 0 nodes busy or unavailable
2019-07-11 15:09:39,903 INFO [jobwatcher:_poll_scheduler_status] 3 nodes required, 3 nodes in asg. Noop
2019-07-11 15:10:40,188 INFO [utils:get_asg_settings] min/desired/max 0/3/10
2019-07-11 15:10:40,208 INFO [jobwatcher:_poll_scheduler_status] 3 nodes requested, 0 nodes busy or unavailable
2019-07-11 15:10:40,208 INFO [jobwatcher:_poll_scheduler_status] 3 nodes required, 3 nodes in asg. Noop
2019-07-11 15:11:40,504 INFO [utils:get_asg_settings] min/desired/max 0/0/10
2019-07-11 15:11:40,524 INFO [jobwatcher:_poll_scheduler_status] 3 nodes requested, 0 nodes busy or unavailable
2019-07-11 15:11:40,524 INFO [jobwatcher:_poll_scheduler_status] Setting desired to 3 nodes, requesting 3 more nodes from asg.

```

`/var/log/sqswatcher`:
```
2019-07-11 15:05:36,065 INFO [sqswatcher:main] sqswatcher startup
2019-07-11 15:05:36,065 INFO [sqswatcher:_get_config] Reading /etc/sqswatcher.cfg
2019-07-11 15:05:36,071 INFO [sqswatcher:_get_config] Configured parameters: region=us-east-2 scheduler=sge sqsqueue=parallelcluster-evcl-SQS-W1YWQ79JCLMP table_name=parallelcluster-evcl-DynamoDBTable-17YFYSQKQ8N5U cluster_user=ec2-user proxy=NONE stack_name=parallelcluster-evcl
2019-07-11 15:05:36,725 WARNING [utils:get_asg_name] Unable to get ASG for stack parallelcluster-evcl
2019-07-11 15:05:56,955 WARNING [utils:get_asg_name] Unable to get ASG for stack parallelcluster-evcl
2019-07-11 15:06:37,162 INFO [utils:get_asg_name] ASG parallelcluster-evcl-ComputeFleet-18TYEPZ0QJEPK found for the stack parallelcluster-evcl
2019-07-11 15:06:38,216 INFO [sqswatcher:_poll_queue] Refreshing cluster properties
2019-07-11 15:06:38,444 INFO [utils:get_asg_settings] min/desired/max 0/3/10
2019-07-11 15:06:38,605 INFO [utils:_get_vcpus_by_instance_type] Instance c5n.large has 2 vcpus.
2019-07-11 15:06:38,606 INFO [utils:_read_cfnconfig] Reading /opt/parallelcluster/cfnconfig
2019-07-11 15:06:38,606 INFO [utils:get_instance_properties] Instance c5n.large will use number of vcpus as slots based on configuration.
2019-07-11 15:06:38,606 INFO [utils:get_instance_properties] Number of slots computed for instance c5n.large: 2
2019-07-11 15:06:38,607 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:06:40,669 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 1 messages from SQS queue
2019-07-11 15:06:40,669 INFO [sqswatcher:_parse_sqs_messages] Unsupported event type autoscaling:TEST_NOTIFICATION. Discarding message.
2019-07-11 15:06:40,670 WARNING [sqswatcher:_parse_sqs_messages] Discarding message sqs.Message(queue_url='https://us-east-2.queue.amazonaws.com/436744843580/parallelcluster-evcl-SQS-W1YWQ79JCLMP', receipt_handle='AQEBAiBSidEP0jNHsu7KqAXZewHwhIO5ic/wd1kTp/YVGBPnBY97Ogbp2UWo2k5pydmq17J10ulzNrcdT984gOFNzzSqjp/bJhWXTDO7rjDq3tvFfhHMF9O/mq6y6/Tmag4usFw8spqEuDAAz5NUuef08HEZzDuS+z6vFK/DRU8+fmqBRC/CAP0ze7OkHhtXZ+3w5n5JBhlqZ9faOuQQbIigd5W7fP59lZHpgXvzBPfC2WOJXgsW8WEP8998uv1EXiB432vsfiokVwcBM4iqH2Vlody8MZPmVJYq9j5UYqqrTPRbAinXpS7uZFtcwzs65sDoYItCy050UYvrEAm68bRAGU+JvlcPczrie6teemTiIwlYk4VJfNYNfZLnUUikDMm+lqSEf8+PpFoHshK3574dbs2I1a3kKBKb4WPDRjcp+y8=')
2019-07-11 15:07:10,703 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:07:12,711 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 0 messages from SQS queue
2019-07-11 15:07:42,741 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:07:44,750 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 0 messages from SQS queue
2019-07-11 15:08:14,780 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:08:16,914 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 3 messages from SQS queue
2019-07-11 15:08:16,914 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-0bc87e8ef9f43917c
2019-07-11 15:08:16,914 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-070121e7d49dbb6fd
2019-07-11 15:08:16,915 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-08d73b053fef1ec75
2019-07-11 15:08:16,915 INFO [sge_commands:exec_qconf_command] Executing operation 'add administrative hosts' for hosts ip-172-31-5-50,ip-172-31-6-200,ip-172-31-0-119
2019-07-11 15:08:16,924 INFO [sge_commands:exec_qconf_command] Executing operation 'add submit hosts' for hosts ip-172-31-5-50,ip-172-31-6-200,ip-172-31-0-119
2019-07-11 15:08:16,963 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:08:16,964 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:08:16,965 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:08:17,019 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:08:17,041 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-5-50: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:08:17,055 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:08:17,056 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:08:17,058 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-0-119: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:08:17,088 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-6-200: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:08:27,302 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-0-119 with error Timeout occurred when executing remote command
2019-07-11 15:08:27,316 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-5-50 with error Timeout occurred when executing remote command
2019-07-11 15:08:27,331 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-6-200 with error Timeout occurred when executing remote command
2019-07-11 15:08:27,469 INFO [sge_commands:add_hosts_to_group] Adding  to @allhosts group
2019-07-11 15:08:27,470 INFO [sge_commands:add_host_slots] Adding  to all.q queue
2019-07-11 15:08:27,512 WARNING [sqswatcher:_requeue_message] Re-queuing failed event {"Message": "{\"StatusCode\":\"Complete\",\"Description\":\"Succesfully launched i-0bc87e8ef9f43917c\",\"Event\":\"parallelcluster:COMPUTE_READY\",\"EC2InstanceId\":\"i-0bc87e8ef9f43917c\",\"Slots\":\"2\",\"LocalHostname\":\"ip-172-31-5-50.us-east-2.compute.internal\"}", "Type": "Notification", "TTL": 2}
2019-07-11 15:08:27,519 WARNING [sqswatcher:_requeue_message] Re-queuing failed event {"Message": "{\"StatusCode\":\"Complete\",\"Description\":\"Succesfully launched i-070121e7d49dbb6fd\",\"Event\":\"parallelcluster:COMPUTE_READY\",\"EC2InstanceId\":\"i-070121e7d49dbb6fd\",\"Slots\":\"2\",\"LocalHostname\":\"ip-172-31-6-200.us-east-2.compute.internal\"}", "Type": "Notification", "TTL": 2}
2019-07-11 15:08:27,529 WARNING [sqswatcher:_requeue_message] Re-queuing failed event {"Message": "{\"StatusCode\":\"Complete\",\"Description\":\"Succesfully launched i-08d73b053fef1ec75\",\"Event\":\"parallelcluster:COMPUTE_READY\",\"EC2InstanceId\":\"i-08d73b053fef1ec75\",\"Slots\":\"2\",\"LocalHostname\":\"ip-172-31-0-119.us-east-2.compute.internal\"}", "Type": "Notification", "TTL": 2}
2019-07-11 15:08:57,582 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:08:59,591 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 0 messages from SQS queue
2019-07-11 15:09:29,621 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:09:31,793 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 3 messages from SQS queue
2019-07-11 15:09:31,793 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-0bc87e8ef9f43917c
2019-07-11 15:09:31,793 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-070121e7d49dbb6fd
2019-07-11 15:09:31,793 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-08d73b053fef1ec75
2019-07-11 15:09:31,793 INFO [sge_commands:exec_qconf_command] Executing operation 'add administrative hosts' for hosts ip-172-31-5-50,ip-172-31-6-200,ip-172-31-0-119
2019-07-11 15:09:31,804 INFO [sge_commands:exec_qconf_command] Executing operation 'add submit hosts' for hosts ip-172-31-5-50,ip-172-31-6-200,ip-172-31-0-119
2019-07-11 15:09:31,874 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:09:31,877 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:09:31,880 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:09:31,960 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:09:31,967 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-5-50: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:09:31,968 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:09:31,968 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:09:31,970 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-0-119: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:09:31,974 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-6-200: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:09:42,179 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-5-50 with error Timeout occurred when executing remote command
2019-07-11 15:09:42,182 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-0-119 with error Timeout occurred when executing remote command
2019-07-11 15:09:42,186 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-6-200 with error Timeout occurred when executing remote command
2019-07-11 15:09:42,283 INFO [sge_commands:add_hosts_to_group] Adding  to @allhosts group
2019-07-11 15:09:42,283 INFO [sge_commands:add_host_slots] Adding  to all.q queue
2019-07-11 15:09:42,319 WARNING [sqswatcher:_requeue_message] Re-queuing failed event {"Message": "{\"StatusCode\":\"Complete\",\"Description\":\"Succesfully launched i-0bc87e8ef9f43917c\",\"Event\":\"parallelcluster:COMPUTE_READY\",\"EC2InstanceId\":\"i-0bc87e8ef9f43917c\",\"Slots\":\"2\",\"LocalHostname\":\"ip-172-31-5-50.us-east-2.compute.internal\"}", "Type": "Notification", "TTL": 1}
2019-07-11 15:09:42,330 WARNING [sqswatcher:_requeue_message] Re-queuing failed event {"Message": "{\"StatusCode\":\"Complete\",\"Description\":\"Succesfully launched i-070121e7d49dbb6fd\",\"Event\":\"parallelcluster:COMPUTE_READY\",\"EC2InstanceId\":\"i-070121e7d49dbb6fd\",\"Slots\":\"2\",\"LocalHostname\":\"ip-172-31-6-200.us-east-2.compute.internal\"}", "Type": "Notification", "TTL": 1}
2019-07-11 15:09:42,337 WARNING [sqswatcher:_requeue_message] Re-queuing failed event {"Message": "{\"StatusCode\":\"Complete\",\"Description\":\"Succesfully launched i-08d73b053fef1ec75\",\"Event\":\"parallelcluster:COMPUTE_READY\",\"EC2InstanceId\":\"i-08d73b053fef1ec75\",\"Slots\":\"2\",\"LocalHostname\":\"ip-172-31-0-119.us-east-2.compute.internal\"}", "Type": "Notification", "TTL": 1}
2019-07-11 15:10:12,358 INFO [sqswatcher:_poll_queue] Refreshing cluster properties
2019-07-11 15:10:12,561 INFO [utils:get_asg_settings] min/desired/max 0/3/10
2019-07-11 15:10:12,615 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:10:14,623 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 0 messages from SQS queue
2019-07-11 15:10:44,653 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:10:46,760 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 3 messages from SQS queue
2019-07-11 15:10:46,761 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-070121e7d49dbb6fd
2019-07-11 15:10:46,761 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-0bc87e8ef9f43917c
2019-07-11 15:10:46,761 INFO [sqswatcher:_parse_sqs_messages] Processing COMPUTE_READY event for instance i-08d73b053fef1ec75
2019-07-11 15:10:46,761 INFO [sge_commands:exec_qconf_command] Executing operation 'add administrative hosts' for hosts ip-172-31-6-200,ip-172-31-5-50,ip-172-31-0-119
2019-07-11 15:10:46,772 INFO [sge_commands:exec_qconf_command] Executing operation 'add submit hosts' for hosts ip-172-31-6-200,ip-172-31-5-50,ip-172-31-0-119
2019-07-11 15:10:46,845 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:10:46,853 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:10:46,855 INFO [transport:_log] Connected (version 2.0, client OpenSSH_7.2p2)
2019-07-11 15:10:46,931 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:10:46,939 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-6-200: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:10:46,939 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:10:46,943 INFO [transport:_log] Authentication (publickey) successful!
2019-07-11 15:10:46,945 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-5-50: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:10:46,948 INFO [remote_command_executor:run_remote_command] Executing remote command command on ec2-user@ip-172-31-0-119: sudo sh -c 'cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf'
2019-07-11 15:10:57,151 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-6-200 with error Timeout occurred when executing remote command
2019-07-11 15:10:57,157 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-5-50 with error Timeout occurred when executing remote command
2019-07-11 15:10:57,160 ERROR [remote_command_executor:_pickable_run_remote_command] Failed when executing remote command on node ip-172-31-0-119 with error Timeout occurred when executing remote command
2019-07-11 15:10:57,254 INFO [sge_commands:add_hosts_to_group] Adding  to @allhosts group
2019-07-11 15:10:57,255 INFO [sge_commands:add_host_slots] Adding  to all.q queue
2019-07-11 15:10:57,294 WARNING [sqswatcher:_requeue_message] Discarding the following message since exceeded the number of max retries (2): {u'Message': u'{"StatusCode":"Complete","Description":"Succesfully launched i-070121e7d49dbb6fd","Event":"parallelcluster:COMPUTE_READY","EC2InstanceId":"i-070121e7d49dbb6fd","Slots":"2","LocalHostname":"ip-172-31-6-200.us-east-2.compute.internal"}', u'Type': u'Notification', u'TTL': 0}
2019-07-11 15:10:57,294 WARNING [sqswatcher:_requeue_message] Discarding the following message since exceeded the number of max retries (2): {u'Message': u'{"StatusCode":"Complete","Description":"Succesfully launched i-0bc87e8ef9f43917c","Event":"parallelcluster:COMPUTE_READY","EC2InstanceId":"i-0bc87e8ef9f43917c","Slots":"2","LocalHostname":"ip-172-31-5-50.us-east-2.compute.internal"}', u'Type': u'Notification', u'TTL': 0}
2019-07-11 15:10:57,294 WARNING [sqswatcher:_requeue_message] Discarding the following message since exceeded the number of max retries (2): {u'Message': u'{"StatusCode":"Complete","Description":"Succesfully launched i-08d73b053fef1ec75","Event":"parallelcluster:COMPUTE_READY","EC2InstanceId":"i-08d73b053fef1ec75","Slots":"2","LocalHostname":"ip-172-31-0-119.us-east-2.compute.internal"}', u'Type': u'Notification', u'TTL': 0}
2019-07-11 15:11:27,312 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieving messages from SQS queue
2019-07-11 15:11:29,320 INFO [sqswatcher:_retrieve_all_sqs_messages] Retrieved 0 messages from SQS queue
```

**Comment by lukeseawalker** [comment]
Could you please send the files inside `/opt/sge/default/common/install_logs` folder?

**Comment by artofbeinghuman** [comment]
Sure, on master:
```
ubuntu@ip-172-31-6-226:/shared$ ll /opt/sge/default/common/install_logs
total 12
drwxr-xr-x 2 sgeadmin sgeadmin 4096 Jul 11 17:12 ./
drwxr-xr-x 4 sgeadmin sgeadmin 4096 Jul 11 17:12 ../
-rw-r--r-- 1 sgeadmin sgeadmin 2247 Jul 11 17:12 qmaster_install_ip-172-31-6-226_2019-07-11_17:12:12.log
ubuntu@ip-172-31-6-226:/shared$ cat /opt/sge/default/common/install_logs/qmaster_install_ip-172-31-6-226_2019-07-11_17\:12\:12.log 
Starting qmaster installation!

Installing Grid Engine as user >sgeadmin<



Your $SGE_ROOT directory: /opt/sge

Using SGE_QMASTER_PORT >6444<.

Using SGE_EXECD_PORT >6445<.

Using >default< as CELL_NAME.


Your $SGE_CLUSTER_NAME: p6444

Using >/opt/sge/default/spool/qmaster< as QMASTER_SPOOL_DIR.




Verifying and setting file permissions and owner in >bin<
Verifying and setting file permissions and owner in >examples<
Verifying and setting file permissions and owner in >inst_sge<
Verifying and setting file permissions and owner in >install_execd<
Verifying and setting file permissions and owner in >install_qmaster<
Verifying and setting file permissions and owner in >lib<
Verifying and setting file permissions and owner in >mpi<
Verifying and setting file permissions and owner in >pvm<
Verifying and setting file permissions and owner in >util<
Verifying and setting file permissions and owner in >utilbin<
Verifying and setting file permissions and owner in >doc<
Verifying and setting file permissions and owner in >include<
Verifying and setting file permissions and owner in >man<
Verifying and setting file permissions and owner in >hadoop<
Verifying and setting file permissions and owner in >dtrace<
Verifying and setting file permissions and owner in >qmon<
Verifying and setting file permissions and owner in >start_gui_installer<

Your file permissions were set


Using >true< as IGNORE_FQDN_DEFAULT.
If it is >true<, the domain name will be ignored.


Making directories

Setting spooling method to dynamic
Dumping bootstrapping information
Initializing spooling database


Using >20000-21000< as gid range.
Using >/opt/sge/default/spool< as EXECD_SPOOL_DIR.
Using >root< as ADMIN_MAIL.
Adding default parallel environments (PE)


cp /opt/sge/default/common/sgemaster /etc/init.d/sgemaster.p6444
/usr/sbin/update-rc.d sgemaster.p6444

   Starting Grid Engine qmaster


Creating the default <all.q> queue and <allhosts> hostgroup
root@ip-172-31-6-226.us-east-2.compute.internal added "@allhosts" to host group list
root@ip-172-31-6-226.us-east-2.compute.internal added "all.q" to cluster queue list


Setting scheduler configuration to >Normal< setting! 
changed scheduler configuration
sge_qmaster successfully installed!
```

on compute:
```
ubuntu@ip-172-31-0-155:~$ ll /opt/sge/default/common/install_logs
total 12
drwxr-xr-x 2 sgeadmin sgeadmin 4096 Jul 11 17:12 ./
drwxr-xr-x 4 sgeadmin sgeadmin 4096 Jul 11 17:12 ../
-rw-r--r-- 1 sgeadmin sgeadmin 2247 Jul 11 17:12 qmaster_install_ip-172-31-6-226_2019-07-11_17:12:12.log
ubuntu@ip-172-31-0-155:~$ cat /opt/sge/default/common/install_logs/qmaster_install_ip-172-31-6-226_2019-07-11_17\:12\:12.log 
Starting qmaster installation!

Installing Grid Engine as user >sgeadmin<



Your $SGE_ROOT directory: /opt/sge

Using SGE_QMASTER_PORT >6444<.

Using SGE_EXECD_PORT >6445<.

Using >default< as CELL_NAME.


Your $SGE_CLUSTER_NAME: p6444

Using >/opt/sge/default/spool/qmaster< as QMASTER_SPOOL_DIR.




Verifying and setting file permissions and owner in >bin<
Verifying and setting file permissions and owner in >examples<
Verifying and setting file permissions and owner in >inst_sge<
Verifying and setting file permissions and owner in >install_execd<
Verifying and setting file permissions and owner in >install_qmaster<
Verifying and setting file permissions and owner in >lib<
Verifying and setting file permissions and owner in >mpi<
Verifying and setting file permissions and owner in >pvm<
Verifying and setting file permissions and owner in >util<
Verifying and setting file permissions and owner in >utilbin<
Verifying and setting file permissions and owner in >doc<
Verifying and setting file permissions and owner in >include<
Verifying and setting file permissions and owner in >man<
Verifying and setting file permissions and owner in >hadoop<
Verifying and setting file permissions and owner in >dtrace<
Verifying and setting file permissions and owner in >qmon<
Verifying and setting file permissions and owner in >start_gui_installer<

Your file permissions were set


Using >true< as IGNORE_FQDN_DEFAULT.
If it is >true<, the domain name will be ignored.


Making directories

Setting spooling method to dynamic
Dumping bootstrapping information
Initializing spooling database


Using >20000-21000< as gid range.
Using >/opt/sge/default/spool< as EXECD_SPOOL_DIR.
Using >root< as ADMIN_MAIL.
Adding default parallel environments (PE)


cp /opt/sge/default/common/sgemaster /etc/init.d/sgemaster.p6444
/usr/sbin/update-rc.d sgemaster.p6444

   Starting Grid Engine qmaster


Creating the default <all.q> queue and <allhosts> hostgroup
root@ip-172-31-6-226.us-east-2.compute.internal added "@allhosts" to host group list
root@ip-172-31-6-226.us-east-2.compute.internal added "all.q" to cluster queue list


Setting scheduler configuration to >Normal< setting! 
changed scheduler configuration
sge_qmaster successfully installed!

```

**Comment by artofbeinghuman** [comment]
Hey Luca,
I was wondering since your first response, when you said:
> I have installed mpi4py in all the nodes (master and computes with) `sudo pip install mpi4py`

How exactly did you run the command on the compute nodes? I can log into the master node via `pcluster ssh mycluster` and then run commands, but I expect them to only be executed on this exact master instance. If I want to ssh into the compute nodes currently, I have to add a network security group to the respective compute instance in the EC2 Console, that lets me connect via ssh to this instance, which is blocked by the standard pcluster compute node security group. Afterwards, I can ssh into the compute instance and execute commands. But doing this for all the compute instances cannot seriously be the proposed workflow. 
So how did you execute the simple `pip` command across all instances?

Thanks for you help so far,
Marvin

**Comment by lukeseawalker** [comment]
For a reason we need to understand, the master is not able to SSH into the compute node and execute the command `cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf` as root.

Could you please send at least one of the archive inside the  `/home/logs/compute` from the master node?

> So how did you execute the simple pip command across all instances?
Using a post_install script or using a custom AMI, and your are already using the latter, right?

> How exactly did you run the command on the compute nodes?
Once you are in the master, you can connect to a compute node just executing `ssh <compute private ip>`. You can retrieve the private IP of the compute node from the AWS EC2 console.

**Comment by artofbeinghuman** [comment]
> Using a post_install script or using a custom AMI, and your are already using the latter, right?

Yes, I am using the config mentioned [here](https://github.com/aws/aws-parallelcluster/issues/1187#issuecomment-510487226). But if all of this doesn't work, I am thinking of using python3.5 in the cluster. This way I wouldn't have compile python3.7, which takes several minutes, and thus using a post_install script becomes feasable.

Here are is an archive from `/home/logs/compute` on the master node:
```
localmachine$ tar xvf i-02a3d3d2f495ea2c4.tar.gz 
var/log/
var/log/alternatives.log
var/log/cfn-init.log
var/log/wtmp
var/log/dpkg.log
var/log/apt/
var/log/apt/history.log
var/log/apt/term.log
var/log/fsck/
var/log/munge/
var/log/cfn-wire.log
var/log/cloud-init.log
var/log/cfn-init-cmd.log
var/log/kern.log
var/log/syslog
var/log/amazon/
var/log/amazon/ssm/
var/log/amazon/ssm/hibernate.log
var/log/amazon/ssm/amazon-ssm-agent.log
var/log/amazon/ssm/errors.log
var/log/lxd/
var/log/auth.log
var/log/unattended-upgrades/
var/log/unattended-upgrades/unattended-upgrades-shutdown.log
var/log/lastlog
var/log/supervisord.log
var/log/cloud-init-output.log
var/log/dist-upgrade/
var/log/nodewatcher
var/log/apache2/
var/log/apache2/error.log
var/log/apache2/access.log
var/log/apache2/other_vhosts_access.log
var/log/btmp
```
I'll send you the archive via email.

**Comment by lukeseawalker** [comment]
No relevant error in the logs.

Could you please do the following test for me?
Look for a running compute node in the AWS EC2 console and get the private IP.
Then connect to the master host and do `ssh <compute private IP>`.
Once you are in the compute node do 
```
sudo su -
cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf
```

What's the output of this command?

**Comment by artofbeinghuman** [comment]
ssh'ing into all three compute nodes fails:
```
ubuntu@ip-172-31-9-116:~$ ssh 172.31.0.153
The authenticity of host '172.31.0.153 (172.31.0.153)' can't be established.
ECDSA key fingerprint is SHA256:VHonXIw0vbqihTZGiYwjmucpQtWXmFVMvAycO8rjkhM.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '172.31.0.153' (ECDSA) to the list of known hosts.
Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
ubuntu@ip-172-31-9-116:~$ ssh 172.31.5.110
The authenticity of host '172.31.5.110 (172.31.5.110)' can't be established.
ECDSA key fingerprint is SHA256:rn8YxqKW87WpRdcpZ20JMnQWAEpBh0n0NJYObRRRwbI.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '172.31.5.110' (ECDSA) to the list of known hosts.
Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
ubuntu@ip-172-31-9-116:~$ ssh 172.31.5.186
The authenticity of host '172.31.5.186 (172.31.5.186)' can't be established.
ECDSA key fingerprint is SHA256:PK9X+ZgIi/eRNpr2aciBmt+6VbaRlziKQKafntlW1w8.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '172.31.5.186' (ECDSA) to the list of known hosts.
Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
```

maybe the private key setup between master and compute instances doesn't work as expected.

**Comment by lukeseawalker** [comment]
It's not AutoScalingGroup terminating the instances, but the instances themselves because they see they are not connected to the scheduler.
They are not connected to the cluster because the master node wasn't able to execute the command inst_sge on them.

Can you please run `df -h` and `cat /etc/fstab` in ip-172-31-10-106 and send the /var/log/cfn-init.log if this host?

**Comment by artofbeinghuman** [comment]
Okay, thanks for clearing that up.

I deleted the old cluster. To be clear where this output comes from, let me detail the steps. I started the cluster with this config (only showing the relevant part)
```
[cluster evo]
vpc_settings = public
key_name = **
custom_ami = ami-0882fe6951fe9a741  
scaling_settings = custom
initial_queue_size = 3
max_queue_size = 10
maintain_initial_size = true
scheduler = sge
cluster_type = spot
compute_instance_type = c5n.large
master_instance_type = c5n.large
placement_group = DYNAMIC
placement = compute 
```
then added another security network group in the AWS EC2 console to one of the compute instances, why I normally use to access my instances from my laptop. In the compute instance I execute:
```
ubuntu@ip-172-31-14-75:~$ sudo su -
root@ip-172-31-14-75:~# cd /opt/sge && /opt/sge/inst_sge -noremote -x -auto /opt/parallelcluster/templates/sge/sge_inst.conf
Reading configuration from file /opt/parallelcluster/templates/sge/sge_inst.conf
Install log can be found in: /opt/sge/default/common/install_logs/execd_install_ip-172-31-14-75_2019-07-12_15:30:48.log
root@ip-172-31-14-75:/opt/sge# df -h
Filesystem                                           Size  Used Avail Use% Mounted on
udev                                                 2.5G     0  2.5G   0% /dev
tmpfs                                                498M  7.1M  491M   2% /run
/dev/nvme0n1p1                                        17G   11G  5.8G  66% /
tmpfs                                                2.5G     0  2.5G   0% /dev/shm
tmpfs                                                5.0M     0  5.0M   0% /run/lock
tmpfs                                                2.5G     0  2.5G   0% /sys/fs/cgroup
/dev/loop1                                            89M   89M     0 100% /snap/core/6964
/dev/loop0                                            18M   18M     0 100% /snap/amazon-ssm-agent/1335
/dev/loop2                                            89M   89M     0 100% /snap/core/7270
ip-172-31-9-174.us-east-2.compute.internal:/home      17G   11G  5.8G  66% /home
ip-172-31-9-174.us-east-2.compute.internal:/shared    20G   44M   19G   1% /shared
ip-172-31-9-174.us-east-2.compute.internal:/opt/sge   17G   11G  5.8G  66% /opt/sge
tmpfs                                                498M     0  498M   0% /run/user/1000
root@ip-172-31-14-75:/opt/sge# cat /etc/fstab
LABEL=cloudimg-rootfs	/	 ext4	defaults,discard	0 0
ip-172-31-9-174.us-east-2.compute.internal:/home /home nfs hard,intr,noatime,vers=3,_netdev 0 2
ip-172-31-9-174.us-east-2.compute.internal:/shared /shared nfs hard,intr,noatime,vers=3,_netdev 0 2
ip-172-31-9-174.us-east-2.compute.internal:/opt/sge /opt/sge nfs hard,intr,noatime,vers=3,_netdev 0 2
```

**Comment by artofbeinghuman** [comment]
Alright, I tried to create a cluster with a post_install script now, but it fails.
My pcluster config:
```
[aws]
aws_region_name = us-east-2

[global]
cluster_template = evo
update_check = true
sanity_check = true

[aliases]
ssh = ssh -i ~/.ssh/**.pem ubuntu@{MASTER_IP} {ARGS}


[vpc public]
master_subnet_id = subnet-***
vpc_id = vpc-***

[scaling custom]
scaledown_idletime = 15

[cluster evo]
vpc_settings = public
key_name = **
base_os = ubuntu1604
post_install = s3://**mybucket**/postinstall.sh 
scaling_settings = custom
initial_queue_size = 3
max_queue_size = 10
maintain_initial_size = true
scheduler = sge
cluster_type = spot
compute_instance_type = c5n.large
master_instance_type = c5n.large
placement_group = DYNAMIC
placement = compute 
```
`postinstall.sh`:
```
#!/bin/bash
# postinstall.sh

sudo apt update && sudo apt-get -y upgrade &&
sudo apt --assume-yes install tmux bc &&
sudo apt update && sudo apt-get -y upgrade &&

sudo apt install -y python3-tk python3-pip &&
sudo apt install -y build-essential checkinstall &&
sudo apt install -y libreadline-gplv2-dev libncursesw5-dev libssl-dev libsqlite3-dev tk-dev libgdbm-dev libc6-dev libbz2-dev libffi-dev &&
sudo apt install -y python3-dev zlib1g-dev libjpeg-dev cmake swig python-pyglet python3-opengl libboost-all-dev libsdl2-dev libosmesa6-dev patchelf ffmpeg xvfb &&

pip3 install --user --upgrade pip setuptools &&
pip3 install --user numpy torch torchvision ipython matplotlib && 
pip3 install --user mpi4py baselines atari_py gym[atari] Pillow click &&
pip3 install --user imageio imageio-ffmpeg scikit-image &&
```

and the error output:
```
$ pcluster create postcl
Beginning cluster creation for cluster: postcl
Creating stack named: parallelcluster-postcl
Status: parallelcluster-postcl - ROLLBACK_IN_PROGRESS                           
Cluster creation failed.  Failed events:
  - AWS::EC2::Instance MasterServer Received FAILURE signal with UniqueId i-0205aac3dfcd46574
```

I am right now executing with `--norollback`. Will be reporting back...

**Comment by lukeseawalker** [comment]
You should remove the `&&` at the end of each line in the post_install script.

**Comment by artofbeinghuman** [comment]
Alright, will do. But why should this be a problem?

**Comment by lukeseawalker** [comment]
We think we have found the problem with the custom AMI.
You were using
```
#base_os = ubuntu1604
# Evo ParallelCluster AMI:
custom_ami = ami-0882fe6951fe9a741  
```
hence the master was trying to connect to the compute with the user ec2-user which doesn't exist in the ubuntu AMI.
You should leave `base_os = ubuntu1604` uncommented

**Comment by artofbeinghuman** [comment]
Okay, I assumed that if I put `base_os = ubuntu1604`, that the `custom_ami` would be ignored. But you're saying that `custom_ami` will be used, but the parallelcluster will know to use `ubuntu` as user?

Btw,  the post_install script failed due to an unrelated pip issue.

**Comment by lukeseawalker** [comment]
Ok, thank you for the update.
